lib/SPIRV/libSPIRV/SPIRVModule.cpp

```diff
@@ -2103,70 +2105,275 @@ static std::string to_string(VersionNumber Version) {
           "incorrect SPIR-V version number " + to_string(MI.SPIRVVersion) +
               " - it conflicts with maximum allowed version which is set to " +
-              to_string(M.getMaximumAllowedSPIRVVersion()))) {
-    M.setInvalid();
+              to_string(MI.getMaximumAllowedSPIRVVersion()))) {
+    MI.setInvalid();
+    return I;
+  }
+
+  SPIRVWord Generator = ReadSPIRVWord(I);
+  if (!ErrorLog.checkError(!I.fail(), SPIRVEC_InvalidModule,
+                           "header parsing error")) {
+    MI.setInvalid();
     return I;
   }
 
-  SPIRVWord Generator = 0;
-  Decoder >> Generator;
   MI.GeneratorId = Generator >> 16;
   MI.GeneratorVer = Generator & 0xFFFF;
 
   // Bound for Id
-  Decoder >> MI.NextId;
+  MI.NextId = ReadSPIRVWord(I);
+  if (!ErrorLog.checkError(!I.fail(), SPIRVEC_InvalidModule,
+                           "header parsing error")) {
+    MI.setInvalid();
+    return I;
+  }
 
-  Decoder >> MI.InstSchema;
-  if (!M.getErrorLog().checkError(MI.InstSchema == SPIRVISCH_Default,
-                                  SPIRVEC_InvalidModule,
-                                  "unsupported instruction schema")) {
-    M.setInvalid();
+  MI.InstSchema = static_cast<SPIRVInstructionSchemaKind>(ReadSPIRVWord(I));
+  if (!ErrorLog.checkError(MI.InstSchema == SPIRVISCH_Default,
+                           SPIRVEC_InvalidModule,
+                           "unsupported instruction schema")) {
+    MI.setInvalid();
     return I;
   }
 
-  while (Decoder.getWordCountAndOpCode() && M.isModuleValid()) {
-    SPIRVEntry *Entry = Decoder.getEntry();
-    if (Entry != nullptr)
-      M.add(Entry);
+  SPIRVWord WordCount = 0;
+  Op OpCode = OpNop;
+  SPIRVEntry *Scope = nullptr;
+  while (true) {
+    WordCount = ReadSPIRVWord(I);
+    if (I.fail()) {
+      SPIRVDBG(spvdbgs() << "[SPIRVDecoder] getWordCountAndOpCode FAIL 0 0\n");
+      break;
+    }
+    std::string RawOp;
+    I >> RawOp;
+    OpCode = getNameMap(OpCode).rmap(RawOp);
+    SPIRVDBG(spvdbgs() << "Read word: W = " << RawOp << " V = " << OpCode
+                       << '\n');
+    if (I.fail()) {
+      SPIRVDBG(spvdbgs() << "[SPIRVDecoder] getWordCountAndOpCode FAIL 0 0\n");
+      break;
+    }
+    SPIRVDBG(spvdbgs() << "[SPIRVDecoder] getWordCountAndOpCode " << WordCount
+                       << " " << OpCodeNameMap::map(OpCode) << '\n');
+    if (!MI.isModuleValid()) {
+      break;
+    }
+
+    SPIRVEntry *Entry =
+        parseAndCreateSPIRVEntry(WordCount, OpCode, Scope, MI, I);
+    if (Entry != nullptr) {
+      MI.add(Entry);
+    }
+    if (I.eof()) {
+      SPIRVDBG(spvdbgs() << "[SPIRVDecoder] getWordCountAndOpCode EOF 0 0\n");
+      break;
+    }
+  }
+
+  MI.resolveUnknownStructFields();
+  return I;
+}
+
+std::istream &SPIRVModuleImpl::parseSPIRV(std::istream &I) {
+  SPIRVModuleImpl &MI = *this;
+  MI.setAutoAddCapability(false);
+  MI.setAutoAddExtensions(false);
+
+  struct [[gnu::packed]] Header {
```

Comment:
I have read both advantages and disadvantages of using packed structs. Given that we do not use packed structs anywhere else, I would recommend to not use it.

Reply:
You must consider that in this case packed struct is a necessity - it's a direct target of memcpy (I.read in this case). If packed wouldn't be used you won't have any guarantee that sizeof(Header) is equal to 20 bytes (struct alignment rules comes into play and filling a struct with hidden members).  
If codebase clarity is a major consideration against `[[gnu::packed]]` please note that in this context `Header` is a helper structure reproducing SPIR-V Header layout which have fixed 20 byte (5 x Word) size. It's scope is local to the function and the symbol isn't visible anywhere else.

Reply:
The [Windows CI build log](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/actions/runs/7845239975/job/21409234150?pr=2316) for this PR contains

> warning C5030: attribute 'gnu::packed' is not recognized

So I guess portability might be a concern here?

Reply:
@svenvh Nice catch. It slipped my mind that the application needs to also be buildable by MSVC. 
There are two alternatives here:
* `#pragma pack` in windows build
* because all header contain 4-byte words we can just use array - which is most portable solution.

It seems that the latter is preferable option - rules for size of arrays in standard are more strict so it will have required number of bytes. This solves portability problem.